### PR TITLE
Windows Support

### PR DESCRIPTION
**Pull request by bltavares**
- [x] Yarn/node scripts working
- [x] Electron starts but fails to start `radicle-proxy`
- [x] Update `radicle-proxy` to latest `radicle-link` commits
- [x] Update `coco` to latest `radicle-link` commits

Pending updates to `radicle-proxy` and `coco`:

-  Merged in #1463 [needs another attempt to update]
-  (likely needs) radicle-dev/radicle-link#515

Patches needed upstream:
- https://github.com/radicle-dev/radicle-link/pull/468
- https://github.com/radicle-dev/twemoji-svg-assets/pull/3

### CODE REVIEW

**Comment by bltavares** [comment]
I've back-ported some of the changes from `radicle-link` on top of the current version used on `radicle-upstream`, and I was able to get things to compile, even tho they don't work properly yet.

image

I should be able to open a PR with only the electron changes needed to copile windows ([branch windows-backport](https://github.com/radicle-dev/radicle-upstream/compare/master...bltavares:backport-windows)) - but it would not yet be usable until `radicle-link` gets updated.

**Comment by bltavares** [comment]
After deleting the keystore, I was able to recreate my credentials and have it running 🥳 
image

**Comment by bltavares** [comment]
now that radicle-link has been updated (congrats!) we can try this again

**Comment by bltavares** [comment]
`yarn run start:dev:win` should run it now ;)

**Comment by bltavares** [comment]
Even tho it runs, it still has windows errors, likely with fixes needed on `radicle-link`

```
 Cloning failed with: fetcher error: invalid path for filesystem 'C:/Users/bltav/AppData/Roaming/radicle/radicle-link/data/17561e66-70b2-429c-8cbe-4d0a978d25bc/git/refs/namespaces/hnrkehankrg4owmaohjxr5gug3u3fpdnd3hxo/refs/remotes/hynkyndc6w3p8urucakobzna7sxwgcqny7xxtw88dtx3pkf7m3nrzc/rad/ids/hnrkqdpm9ub19oc8dccx44echy76hzfsezyio': The filename or extension is too long.
```

Should we merge this or wait for more fixes to make it fully useable?

**Comment by bltavares** [comment]
I was able to circumvent windows long-path limits and get Radicle Upstream to clone projects from the sprout server

```pwsh
$env:RAD_HOME="D:\rad"
yarn run start:dev:win
```

image

I had a few issues starting the project - something about #1496 race condition. It looks like the `POST /v1/identities` has a race condition, and when I added a `console.error` on the onboarding screen, between creating the keystore and fetching the identity, I was able to continue.

I think we are good to go with the `RAD_HOME` pointing to a root folder :)

**Comment by bltavares** [comment]
If someone wants to try [on a Windows VM](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) close to what others might have (i.e. no developer tools) you [can test this binary](https://github.com/bltavares/radicle-upstream/releases/tag/pr-1623)

**Review by rudolfs — CHANGES_REQUESTED** [review]
Really awesome work, Bruno! 🎉 
Can't wait for this to land :)

Left some remarks and questions inline. 

> Should we merge this or wait for more fixes to make it fully usable?

I think if we clearly document it as being experimental, we could merge it in. This way we don't have to put in more work keeping this branch up-to-date.

Here's a couple of things I'd like to clarify before moving on:
  - did you try running the unit and integration tests on windows, if so, how much breakage was there?
  - I think it would be also beneficial to document the next steps to make the windows user experience on-par with what we already have for Linux/macOS (e.g. running tests on a Windows CI, listing all the little things that need to be fixed, etc.)

And lastly, we require all commits to be verified, see [here](https://docs.github.com/en/github/authenticating-to-github/managing-commit-signature-verification) for how to set that up.

File: package.json
Comment (geigerzaehler) [inline package.json:149]: I’m wondering whether the proxy path should be detected by the app. We would only need one script then.

File: package.json
Comment (geigerzaehler) [inline package.json:153]: I’m curious why we don’t use `cross-env` for the second command `TZ=UTC yarn test:integration`.

File: proxy/coco/src/state.rs
Comment (geigerzaehler) [inline proxy/coco/src/state.rs:1152]: Could you explain why you added this? Without the `drop()` Rust should keep `tmp_dir` around until it goes out of scope, i.e. here. Maybe you were running into this error (#1378)?

```
test can_ask_and_clone_project ... fatal: '.' does not appear to be a git repository
```

**Review by geigerzaehler — COMMENTED** [review]
Great effort! Looking forward to having this merged.

**Comment by geigerzaehler** [comment]
I’ve added issue #1847 for Windows CI builds.

File: package.json
Comment (bltavares) [reply package.json:153]: I missed it hehe

File: proxy/coco/src/state.rs
Comment (bltavares) [reply proxy/coco/src/state.rs:1152]: In this case we can remove it. I had a few tests on previously where the tmpdir was moved into a method, and was dropped before finishing the test.

You are right, the method should drop by the end of it. I've added the `drop` as a guardrail to prevent moving the tmpdir and dropping it earlier with a compiler error.

File: package.json
Comment (bltavares) [reply package.json:149]: Yes, I've considered this also, but it was more effort

File: ui/src/profile.ts
Comment (bltavares) [reply ui/src/profile.ts:6]: On VSCode it was complaining about it

File: ui/Screen/Onboarding.svelte
Comment (bltavares) [reply ui/Screen/Onboarding.svelte:54]: The `withRetry` only fails if it gets a `Failed to fetch` message, meaning the connection didn't work. Here, the proxy is returning a `403` and it's not retrying as it has a different error message.

I'm wrapping it to trigger the retry on 403 as well. There is likely a more elegant way to allow retries on different status code but I mostly wanted to get it working.

File: proxy/api/src/config.rs
Comment (bltavares) [reply proxy/api/src/config.rs:51]: Windows users might have set HOME previously - specially on dev envs. It could be done with `cfg` tho.

File: package.json
Comment (bltavares) [reply package.json:229]: I can remove it - no need to keep all those tools on the repo. Volta only runs if this entry exists, so it's a side-effect from my env.

**Comment by bltavares** [comment]
> did you try running the unit and integration tests on windows, if so, how much breakage was there?
Most of the tests are passing. There are a few integration tests failing, likely due to difference between windows/linux concurrent file access and tmpdir location and deletion.

> And lastly, we require all commits to be verified
Do they need to be GPG signed, not only with `signed off`?

File: package.json
Comment (geigerzaehler) [reply package.json:149]: :ok_hand: Would you mind creating an issue for this after the PR is merged?

**Comment by rudolfs** [comment]
> Do they need to be GPG signed, not only with signed off?

Correct, we also [require GPG signing](https://github.com/radicle-dev/radicle-upstream/blob/master/DEVELOPMENT.md#how-do-we-work) in this repo.

File: ui/Screen/Onboarding.svelte
Comment (geigerzaehler) [reply ui/Screen/Onboarding.svelte:54]: This is curious: `createIdentity` should never return a 403 :thinking:. If it does, there is something wrong and we don’t want to retry.  Can you reproduce this @bltavares? If so, we should open a separate issue.

Comment (bltavares) [reply ui/Screen/Onboarding.svelte:54]: It happens frequently, but sometimes it works. There seem to be a race condition on the proxy, where the `session.createKeystore()` respond before the session token is stored, and then `createIdentity` fails to authenticate. #1496 seems to showcase it's happening on other OSes

Comment (geigerzaehler) [reply ui/Screen/Onboarding.svelte:54]: To me it seems like #1496 describes an issue where identity creation fails permanently and requires manual interaction to resolve. If we keep retrying this would just delay showing an error. If it fails once and then recovers we also loose valuable information about existing errors if we keep re-trying. This is why I think we should not include this code and find a solution to the underlying issue.

Could you add error information to #1496. You can get it by clicking “copy to clipboard” when an error shows. Our error reporting infrastructure has improved since the issue was opened.

File: proxy/coco/src/state.rs
Comment (geigerzaehler) [reply proxy/coco/src/state.rs:1152]: Not relevant anymore. The file is gone.

File: package.json
Comment (rudolfs) [reply package.json:45]: I just tried the globs, didn't work.

File: ui/Screen/Onboarding.svelte
Comment (rudolfs) [reply ui/Screen/Onboarding.svelte:54]: I removed this in https://github.com/radicle-dev/radicle-upstream/pull/2375.

File: package.json
Comment (rudolfs) [reply package.json:149]: Done in https://github.com/radicle-dev/radicle-upstream/pull/2375.

File: package.json
Comment (rudolfs) [reply package.json:153]: Fixed in https://github.com/radicle-dev/radicle-upstream/pull/2375.

**Comment by rudolfs** [comment]
Closing this in favour of: https://github.com/radicle-dev/radicle-upstream/pull/2375.

**Pull request closed without merging**
